File: traffic-gen/snowtg.lua
-- Startup-only scenario helpers. The Python launcher runs this bridge once.
assert(math.type, "Lua 5.3 or newer required")
local M = {}

local function checked(options, allowed)
    if options == nil then options = {} end
    assert(type(options) == "table", "options must be a table")
    local names = {}
    for name in allowed:gmatch("%S+") do names[name] = true end
    for key in pairs(options) do
        assert(names[key], "unknown option: " .. tostring(key))
    end
    return options
end

local function default(value, fallback)
    if value == nil then return fallback end
    return value
end

-- Rates count logical transactions (including Keep-Alive reuse), not sockets.
-- Duration is seconds; options.start selects a linear ramp to cps.
function M.phase(name, duration, cps, options)
    options = checked(options, "start")
    return {name=name, duration_sec=duration, target_cps=cps, start_cps=options.start}
end

function M.http(name, ip, port, options)
    options = checked(options, "weight path keepalive method host")
    local keepalive = options.keepalive
    if keepalive == nil then keepalive = false end
    return {name=name, weight=default(options.weight, 1), transport="tcp",
        peer={ip=ip, port=default(port, 80)},
        http={method=default(options.method, "GET"), path=default(options.path, "/"),
              host=options.host, keepalive=keepalive}}
end

function M.dns(name, ip, qname, port, options)
    options = checked(options, "weight qtype")
    return {name=name, weight=default(options.weight, 1), transport="udp",
        peer={ip=ip, port=default(port, 53)}, dns={qname=qname, qtype=default(options.qtype, "A")}}
end

-- Concurrency is global. Use phases OR duration/cps; native validation checks
-- protocol/rate limits after the launcher removes managed-run metadata.
function M.scenario(name, options)
    options = checked(options, "classes concurrency report_interval phases duration cps assertions purpose service seed")
    local result = {name=name, load_model="open", classes=options.classes,
        assertions=options.assertions, purpose=options.purpose, service=options.service, seed=options.seed,
        max_concurrency=default(options.concurrency, 256),
        report_interval_sec=default(options.report_interval, 1)}
    if options.phases ~= nil then
        assert(options.duration == nil and options.cps == nil,
               "use phases OR duration/cps, not both")
        result.phases = options.phases
    else
        assert(options.duration ~= nil and options.cps ~= nil,
               "provide phases or both duration and cps")
        result.duration_sec, result.target_cps = options.duration, options.cps
    end
    return result
end

-- Ratios use [0,1], latency uses ms and quantile in (0,1]. Selectors/thresholds
-- are validated by the managed runner; critical failures yield exit code 2.
function M.assertion(metric, op, value, options)
    options = checked(options, "class_name protocol phase quantile latency_metric critical step")
    return {metric=metric, op=op, value=value, class=options.class_name,
        protocol=options.protocol, phase=options.phase, quantile=options.quantile,
        latency_metric=options.latency_metric, critical=default(options.critical, true), step=options.step}
end

-- Business helpers mirror Python; they build plain tables, never callbacks.
-- Return declarative selectors; values are resolved by the native business owner.
function M.ref(name) return {ref=name} end
-- Header indices are zero-based. JSON paths are Pointers, not Lua expressions.
function M.extract(source, path, options)
    options=checked(options,"index")
    return {source=source,path=default(path,""),index=default(options.index,0)}
end
-- Preserve a false right operand; using "right or default" would change its type/value.
function M.check(left, op, right)
    local result={left=left,op=op}
    if op ~= "exists" then result.right=right end
    return result
end
-- One weighted class owns the entire ordered business. String datasets become file
-- references for launcher expansion; workers never execute Lua or read those files.
function M.transaction(name, options)
    options=checked(options,"steps weight vars dataset timeout_ms")
    local data=options.dataset
    if type(data)=="string" then data={file=data} end
    return {name=name,weight=default(options.weight,1),transaction={steps=options.steps,
        vars=options.vars,dataset=data,timeout_ms=options.timeout_ms}}
end
-- Per-step connection policy, extraction, and checks accompany the protocol fields.
-- Dynamic endpoint/Host/body templates are expanded only in the native context.
function M.http_step(name, ip, port, options)
    options=checked(options,"path method host keepalive headers body extract checks next timeout_ms")
    return {name=name,type="http",peer={ip=ip,port=default(port,80)},
        http={path=default(options.path,"/"),method=default(options.method,"GET"),
              host=options.host,keepalive=default(options.keepalive,false),headers=options.headers,body=options.body},
        extract=options.extract,checks=options.checks,next=options.next,timeout_ms=options.timeout_ms}
end
-- The business DNS path uses A records; export address into ctx for later requests.
function M.dns_step(name, ip, qname, port, options)
    options=checked(options,"extract checks next timeout_ms")
    return {name=name,type="dns",peer={ip=ip,port=default(port,53)},dns={qname=qname,qtype="A"},
        extract=options.extract,checks=options.checks,next=options.next,timeout_ms=options.timeout_ms}
end
-- Pass ms for fixed waits, or nil plus minimum/maximum for deterministic native sampling.
function M.think(name, ms, options)
    options=checked(options,"minimum maximum next")
    return {name=name,type="think",ms=ms,min_ms=options.minimum,max_ms=options.maximum,next=options.next}
end
-- Lua reserves then: callers use {["then"]="step", otherwise="other"}.
-- Both targets must be forward names or end; unselected work is skipped, not failed.
function M.branch(name, condition, options)
    options=checked(options,"then otherwise")
    return {name=name,type="branch",condition=condition,["then"]=options["then"],["else"]=options.otherwise}
end

if ... == "snowtg" then return M end

-- ponytail: serialize only JSON-compatible Lua values; no JSON package needed.
-- Empty tables are arrays; mixed/sparse tables, cycles and non-finite numbers
-- are errors, never silently dropped. Schema validation stays in the C loader.
local function encode(value, seen)
    local kind = type(value)
    if kind == "string" then
        return '"' .. value:gsub('[%z\1-\31\\"]', function(c)
            return string.format('\\u%04x', c:byte())
        end) .. '"'
    elseif kind == "boolean" then
        return tostring(value)
    elseif kind == "number" then
        assert(value == value and value ~= math.huge and value ~= -math.huge,
               "non-finite number in plan")
        if math.type(value) == "integer" then return tostring(value) end
        if value == math.floor(value) then return string.format("%.0f", value) end
        return string.format("%.17g", value)
    end
    assert(kind == "table", "unsupported plan value: " .. kind)
    assert(not seen[value], "cyclic table in plan")
    seen[value] = true
    local keys, array, count = {}, true, 0
    for key in pairs(value) do
        count = count + 1
        keys[count] = key
        if type(key) ~= "number" then array = false end
    end
    local parts = {}
    if array then
        for _, key in ipairs(keys) do
            assert(key >= 1 and key <= count and key == math.floor(key),
                   "sparse/non-sequence table in plan")
        end
        for i = 1, count do parts[i] = encode(value[i], seen) end
    else
        for _, key in ipairs(keys) do
            assert(type(key) == "string", "mixed/non-string keys in plan")
        end
        table.sort(keys)
        for i, key in ipairs(keys) do
            parts[i] = encode(key, seen) .. ":" .. encode(value[key], seen)
        end
    end
    seen[value] = nil
    return (array and "[" or "{") .. table.concat(parts, ",") .. (array and "]" or "}")
end

-- Execute trusted user code once with sibling imports and script-only argv.
-- A returned table takes precedence over the optional global plan variable.
local script, output_path = arg[1], arg[2]
package.loaded.snowtg = M
local directory = assert(script:match("^(.*)/"))
package.path = directory .. "/?.lua;" .. directory .. "/?/init.lua;" .. package.path
arg = {[0]=script}
local result = dofile(script)
if result == nil then result = _G.plan end
assert(type(result) == "table", "script must return a table or export 'plan'")
local text = encode(result, {}) .. "\n"
local output = assert(io.open(output_path, "w"))
assert(output:write(text))
assert(output:close())

File: traffic-gen/snowtg.py
#!/usr/bin/env python3
"""Write a Python/Lua scenario exporting `plan`, then run the native generator.

    python3 traffic-gen/snowtg.py scenario.py -- -l 0,1 -- --workers 1
    python3 traffic-gen/snowtg.py scenario.lua -- -l 0,1 -- --workers 1
    python3 traffic-gen/snowtg.py --emit-json plan.json scenario.py
    python3 traffic-gen/snowtg.py run --output debug/run1 scenario.lua -- <EAL> -- <app>
    python3 traffic-gen/snowtg.py compare baseline/result.json candidate/result.json
    python3 traffic-gen/snowtg.py report run/result.json --output report.html

Scripts may import scenario, phase, http and dns from snowtg. Helpers produce
ordinary dictionaries; Lua scripts use require("snowtg") and may return a plan.
Lua requires lua, lua5.4 or lua5.3 on PATH (or SNOWTG_LUA pointing to it).
The existing C compiler validates the schema at startup.
"""

import argparse
import json
import os
from pathlib import Path
import runpy
import shutil
import subprocess
import sys


def phase(name, duration, cps, *, start=None):
    """Constant or linear-ramp transaction starts/second; duration is seconds.

    A reused Keep-Alive request consumes one arrival, just like a new connection.
    """
    result = dict(name=name, duration_sec=duration, target_cps=cps)
    if start is not None:
        result["start_cps"] = start
    return result


def http(name, ip, port=80, *, weight=1, path="/", keepalive=False,
         method="GET", host=None):
    """One HTTP traffic class; keepalive=True reuses connections."""
    config = dict(method=method, path=path, keepalive=keepalive)
    if host is not None:
        config["host"] = host
    return dict(name=name, weight=weight, transport="tcp",
                peer=dict(ip=ip, port=port), http=config)


def dns(name, ip, qname, port=53, *, weight=1, qtype="A"):
    """One UDP DNS traffic class (A or AAAA query)."""
    return dict(name=name, weight=weight, transport="udp",
                peer=dict(ip=ip, port=port), dns=dict(qname=qname, qtype=qtype))


def ref(name):
    """Describe a typed vars/data/ctx lookup without evaluating it in Python."""
    return {"ref": name}


def extract(source, path="", *, index=0):
    """Select a response scalar for native extraction.

    Sources are DNS address or HTTP status/header/json. Header index is zero-based;
    JSON path is a JSON Pointer. Export happens before protocol parser reset.
    """
    return dict(source=source, path=path, index=index)


def check(left, op, right=None):
    """Build a typed comparison; exists intentionally omits the right operand.

    Keep literal types, including False and None. Native execution rejects
    incompatible operand types instead of implicitly converting strings.
    """
    return dict(left=left, op=op, **({} if op == "exists" else {"right": right}))


def transaction(name, *, steps, weight=1, vars=None, dataset=None, timeout_ms=None):
    """Build a weighted business class with ordered native steps.

    One admission covers the entire business, including think time. Dataset paths
    are expanded relative to the script directory before execution; immutable rows
    cycle by global planned ordinal, including arrivals skipped under overload.
    """
    config = dict(steps=list(steps))
    for key, value in (("vars", vars), ("dataset", dataset), ("timeout_ms", timeout_ms)):
        if value is not None:
            config[key] = {"file": str(value)} if key == "dataset" and isinstance(value, (str, Path)) else value
    return dict(name=name, weight=weight, transaction=config)


def http_step(name, ip, port=80, *, path="/", method="GET", host=None,
              keepalive=False, headers=None, body=None, extract=None, checks=None,
              next=None, timeout_ms=None):
    """Build a network step, without independent class weight or admission.

    Templates are resolved against the active native context. keepalive requests
    reuse only when endpoint/Host and response policy permit it. next names a
    later step or end; headers/body cannot override plugin-owned framing.
    """
    step = http(name, ip, port, path=path, method=method, host=host, keepalive=keepalive)
    step.pop("weight"); step.pop("transport"); step["type"] = "http"
    for key, value in (("headers", headers), ("body", body)):
        if value is not None: step["http"][key] = value
    step.update({k: v for k, v in dict(extract=extract, checks=checks, next=next, timeout_ms=timeout_ms).items() if v is not None})
    return step


def dns_step(name, ip, qname, port=53, *, extract=None, checks=None, next=None, timeout_ms=None):
    """Build an A-query step whose exported address may feed a later HTTP target.

    Extraction and checks run on the complete datagram. The native resolver only
    accepts addresses matching the question or an answer-section CNAME chain.
    """
    step = dns(name, ip, qname, port)
    step.pop("weight"); step.pop("transport"); step["type"] = "dns"
    step.update({k: v for k, v in dict(extract=extract, checks=checks, next=next, timeout_ms=timeout_ms).items() if v is not None})
    return step


def think(name, ms=None, *, minimum=None, maximum=None, next=None):
    """Describe fixed milliseconds or an inclusive minimum/maximum range.

    The native compiler rejects conflicting forms. An owner timer schedules the
    continuation; this helper neither sleeps nor samples a Python random number.
    """
    return dict(name=name, type="think", **{k: v for k, v in
        dict(ms=ms, min_ms=minimum, max_ms=maximum, next=next).items() if v is not None})


def branch(name, condition, *, then, otherwise):
    """Build a forward-only conditional jump using a check/ref description.

    otherwise maps to JSON else. Targets name later steps or end; this is not
    a Python callback, and the unselected path does not count as request failure.
    """
    return dict(name=name, type="branch", condition=condition, then=then, **{"else": otherwise})


def assertion(metric, op, value, *, class_name=None, protocol=None, phase=None,
              quantile=None, latency_metric=None, critical=True, step=None):
    """Build an SLO spec; the managed runner validates selectors and thresholds.

    Rates are fractions, latency_ms is milliseconds with quantile in (0,1].
    Latency defaults to successful completion; critical failures set exit code 2.
    """
    result = dict(metric=metric, op=op, value=value, critical=critical)
    result.update({k: v for k, v in dict(
        **{"class": class_name}, protocol=protocol, phase=phase,
        quantile=quantile, latency_metric=latency_metric, step=step).items() if v is not None})
    return result


def scenario(name, *, classes, concurrency=256, phases=None, duration=None,
             cps=None, report_interval=1, assertions=None, purpose=None, service=None, seed=None):
    """Open arrivals: provide phases OR duration/cps; concurrency is global."""
    result = dict(name=name, load_model="open", max_concurrency=concurrency,
                  report_interval_sec=report_interval, classes=list(classes))
    if phases is not None:
        if duration is not None or cps is not None:
            raise ValueError("use phases OR duration/cps, not both")
        result["phases"] = list(phases)
    else:
        if duration is None or cps is None:
            raise ValueError("provide phases or both duration and cps")
        result.update(duration_sec=duration, target_cps=cps)
    for key, value in (("assertions", assertions), ("purpose", purpose), ("service", service), ("seed", seed)):
        if value is not None:
            result[key] = value
    return result


def load_script(path):
    """Execute a trusted script once, with sibling imports and script-only argv."""
    old_path, old_argv = sys.path[:], sys.argv
    try:
        sys.path.insert(0, str(path.parent))
        sys.argv = [str(path)]
        namespace = runpy.run_path(str(path), run_name="__main__")
    finally:
        sys.path[:] = old_path
        sys.argv = old_argv
    plan = namespace.get("plan")
    if not isinstance(plan, dict):
        raise ValueError(f"{path}: script must export a dict named 'plan'")
    return json.dumps(plan, ensure_ascii=False, allow_nan=False, indent=2) + "\n"


def load_lua(path):
    """Execute trusted Lua once; an inherited Linux memfd carries only its plan.

    Script logs remain on stdout/stderr. This executes ordinary user code with
    the launcher's privileges, as load_script does; neither is a sandbox.
    """
    interpreter = os.environ.get("SNOWTG_LUA") or next(
        (found for name in ("lua", "lua5.4", "lua5.3")
         if (found := shutil.which(name))), None)
    if not interpreter:
        raise ValueError("Lua interpreter missing: install lua5.4 or set SNOWTG_LUA")
    # Separate descriptor keeps print()/io.write() logs out of the JSON data.
    with os.fdopen(os.memfd_create("snowtg-lua"), "r+b") as output:
        result = subprocess.run([interpreter, str(Path(__file__).with_suffix(".lua")),
                                 str(path), f"/proc/self/fd/{output.fileno()}"],
                                pass_fds=(output.fileno(),))
        if result.returncode:
            raise ValueError(f"Lua scenario failed (exit {result.returncode}): {path}")
        output.seek(0)
        text = output.read().decode("utf-8")
        if not isinstance(json.loads(text), dict):
            raise ValueError("Lua scenario must produce a plan object")
        return text


def main(argv=None):
    argv = list(sys.argv[1:] if argv is None else argv)
    if argv[:1] in (["compare"], ["report"]):
        from snowtg_results import cli
        try:
            return cli(argv)
        except (OSError, ValueError, KeyError, TypeError) as error:
            print(f"snowtg: {error}", file=sys.stderr)
            return 1
    managed = argv[:1] == ["run"]
    if managed:
        argv = argv[1:]
    parser = argparse.ArgumentParser(description=__doc__,
        formatter_class=argparse.RawDescriptionHelpFormatter)
    parser.add_argument("--emit-json", type=Path, metavar="PATH",
                        help="export only; native schema validation happens when run")
    parser.add_argument("--output", type=Path, help="new managed-run directory (CSV, result.json, report.html)")
    parser.add_argument("--baseline", type=Path, help="baseline result.json for the report")
    parser.add_argument("--timeout", type=float, help="managed run wall-clock limit in seconds")
    parser.add_argument("--binary", type=Path,
                        default=Path(__file__).resolve().parent / "build/traffic-gen",
                        help="native executable (default: adjacent build/traffic-gen)")
    parser.add_argument("script", type=Path, help="trusted .py/.lua script or existing .json")
    parser.add_argument("args", nargs=argparse.REMAINDER,
                        help="EAL arguments, then -- and traffic-gen arguments")
    options = parser.parse_args(argv)
    try:
        path = options.script.resolve()
        if path.suffix not in (".py", ".lua", ".json"):
            raise ValueError("scenario must be a .py, .lua or .json file")
        if path.suffix == ".lua":
            text = load_lua(path)
        else:
            text = load_script(path) if path.suffix == ".py" else path.read_text()
        from snowtg_datasets import expand_datasets
        plan = json.loads(text)
        if not isinstance(plan, dict): raise ValueError("scenario must be an object")
        if expand_datasets(plan, path.parent):
            text = json.dumps(plan, ensure_ascii=False, allow_nan=False, indent=2) + "\n"
        if options.emit_json:
            if managed or options.output or options.baseline or options.timeout is not None:
                raise ValueError("--emit-json cannot be combined with managed-run options")
            if options.args:
                raise ValueError("--emit-json does not accept runtime arguments")
            if options.emit_json.resolve() == path:
                raise ValueError("output must not overwrite the input scenario")
            options.emit_json.write_text(text, encoding="utf-8")
            return 0
        plan = json.loads(text)
        if not isinstance(plan, dict):
            raise ValueError("scenario must be an object")
        if managed or options.output or options.baseline or options.timeout is not None or any(
                key in plan for key in ("assertions", "purpose", "service", "dataset_sources")):
            # Managed mode waits for the native child and evaluates final CSVs;
            # scripts still execute only at startup, never in the packet path.
            from snowtg_results import run
            return run(plan, options.binary, options.args, options.output, options.timeout, options.baseline)
        args = options.args
        if args[:1] == ["--"]:
            args = args[1:]
        if "--" not in args:
            args = [*args, "--"]
        # ponytail: Linux memfd lets exec keep the JSON without a temp file or
        # resident Python process; the C loader already supports seekable files.
        fd = os.memfd_create("snowtg-scenario", flags=0)
        try:
            with os.fdopen(os.dup(fd), "w", encoding="utf-8") as output:
                output.write(text)
                output.flush()
                output.seek(0)
            binary = str(options.binary.resolve())
            sys.stdout.flush()
            sys.stderr.flush()
            os.execv(binary, [binary, *args, f"/proc/self/fd/{fd}"])
        finally:
            os.close(fd)
    except (OSError, ValueError, TypeError, SyntaxError, KeyError) as error:
        print(f"snowtg: {error}", file=sys.stderr)
        return 1


if __name__ == "__main__":
    sys.exit(main())
